### ISSUE TITLE
Error running Haskell program

### ISSUE THREAD

**Issue by jnear** [oscarlab/graphene#806]
Hi! I'm trying to run a statically-linked Haskell binary (compiled with GHC) in an SGX enclave using Graphene. This binary runs just fine within Graphene *without* SGX, but we run into problems running with SGX. We get similar issues with *all* GHC programs. Here's an example of a "hello world" program:

```
module Main where

main :: IO ()
main = do
  putStrLn "Hello world!"
```

The output from Graphene is attached below. We get errors like:

```
[     1] ---- shim_getrlimit(9,0x7055b2f0) = 0
[     1] ---- shim_mmap(0x4200000000,1048576,PROT_NONE,MAP_PRIVATE|MAP_ANON|0x4000,-1,0) ... mmap
[     1] bkeep_mmap: 0x4200000000-0x14200100000
*** allocating out of heap: 0x4200000000 ***
[     1] bkeep_munmap: 0x4200000000-0x14200100000
[     1] ---- return from shim_mmap(...) = -1
[     1] ---- shim_mmap(0x0,1048576,PROT_NONE,MAP_PRIVATE|MAP_ANON|0x4000,-1,0) ... mmap
[     1] ---- return from shim_mmap(...) = -12
```

Is there an obvious reason this might be happening, or something we can change to fix it? Alternatively, is there more information we can provide to help in debugging it?

Thanks!
[helloworld-output.txt](https://github.com/oscarlab/graphene/files/3347621/helloworld-output.txt)

**Comment by dimakuv** [comment]
Dear Joseph,

Thanks for the bug report. The culprit here is the weird mmap'ed address `0x4200000000` with size `1048576` (i.e., Haskell runtime wants to allocate 1MB at address 264GB).

Graphene with the SGX backend doesn't support such allocations. The reasons are twofold:
1. Intel SGX hardware by itself limits the maximum enclave size to 64GB so this already precludes correct allocation at address 264GB (and knowing that enclave base address is 0x0).
2. Graphene-SGX software has a `sgx.enclave_size` option in the manifest which is by default is 256MB (see https://github.com/oscarlab/graphene/wiki/SGX-Manifest-Syntax#enclave-size-optional). Even if you will set this option to `sgx.enclave_size = 300G`, the SGX enclave will refuse to run (actually never tried it myself, not sure what kind of error will be reported :)).

There are some explanations as to why Haskell compiler/runtime needs this weird address:
* https://github.com/ghc/ghc/blob/bd660edeb783a74e5ca3f1f82713b2aeedae19dc/rts/posix/OSMem.c#L510 -- comments in source code are especially useful
* https://gitlab.haskell.org/ghc/ghc/issues/10877 -- one relevant issue
* https://phabricator.haskell.org/D2513 -- another relevant issue

In short, Haskell runtime wants to reserve the lower address space for JIT-generated code, and starts its heap at this absurdly-high address to not conflict with the code pages and any other libraries (in particular OpenCL).

*Solution?* To be honest, I don't know how to circumvent this easily. Hopefully there is a configuration option / environment variable / CLI knob to change this address to some lower address? Alternatively, you can go through the source code of Haskell (link above) and find some hack to bypass this.

**Comment by jnear** [comment]
Thank you for the pointers - that's very useful info! We'll dig into the Haskell runtime and see if we can solve this problem. I'll leave a comment in this issue with the solution if we find one. Thanks again!
